### DISCUSSION TITLE
Mac M1 Support : via Colima (Requesting feedback)

### DISCUSSION THREAD

**Discussion by sajeetharan** [Azure/azure-cosmos-db-emulator-docker#114] [Announcements]
We've received many requests on UserVoice and GitHub for M1 Mac support. While we’re working on native support, we’d love feedback on using Colima as a workaround. Some users have already reported success in this repo, but we’d like more input before an official announcement. Here are the steps to try it out:

Here’s a step-by-step guide in markdown format for testing M1 Mac support with Colima as a workaround.

```markdown
# Testing M1 Mac Support with Colima

We've received many requests on UserVoice and GitHub for M1 Mac support. While we’re working on developing native support, we’d appreciate feedback on using Colima as a workaround. Some users have already reported success with this setup, and we’d like more input before making an official announcement. Please follow the steps below to try it out.

## Step 1: Install Colima (https://github.com/abiosoft/colima)

To install Colima, use one of the following commands based on your preferred package manager:

### Using Homebrew

```bash
brew install colima
```

### Using MacPorts

```bash
sudo port install colima
```

### Using Nix

```bash
nix-env -iA nixpkgs.colima
```

## Step 2: Start Colima Docker Builder

After installing Colima, start it with the following command to configure the Docker builder:

```bash
colima start --arch amd
```

## Step 3: Run the Emulator

Once Colima is running, use Docker to run the emulator. Execute the following command:

```bash
docker run --publish 8081:8081 --publish 10250-10255:10250-10255 --name linux-emulator --detach mcr.microsoft.com/cosmosdb/linux/azure-cosmos-emulator:latest
```

## Additional Notes

- If you encounter any issues, please report them in the repository. 
- Your feedback is valuable as it helps us improve and evaluate this workaround for M1 Mac support.

Thank you for your help and feedback!
```

**Comment by mwgustin** [comment]
I tested this out and was able to get the emulator to start up with Colima, but I ran into cert warnings trying to execute locally.  I vaguely remember there being some mac specific commands needed to add the cert to the mac trust store, but I can't seem to dig them up... those should definitely be included in any sort of published instructions.

**Comment by AayeshaShrestha** [comment]
I am able to run the emulator with Colima. It gave self signed certificate error, which was fixed with:
```
curl -k https://localhost:8081/_explorer/emulator.pem -o azure-cosmos-emulator-cert.pem
sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain azure-cosmos-emulator-cert.pem
```

I got the emulator running at `https://localhost:8081/_explorer/index.html` and also connected to Azure Data Studio. 
<img width="1450" alt="Screenshot 2024-11-14 at 1 45 49 PM" src="https://github.com/user-attachments/assets/12da74cf-c507-4f60-a83c-5043e3f8910b">

However the emulator is too slow at times and not running sometimes. I created a database and container and inserted a record in the container. And tried to query the records but it kept spinning for about a minute and threw this error.

<img width="786" alt="Screenshot 2024-11-14 at 1 16 09 PM" src="https://github.com/user-attachments/assets/b4e7cd54-ba46-4db5-93ec-49abd22ff82c">

I was also trying to use azure-cosoms sdk for python. It works till logging the database and container but gets stuck in other methods like create_database, create_container, list_databases. All of them had request timeout issue.

**Comment by atkinsonbg** [comment]
docker run --publish 8081:8081 --publish 10250-10255:10250-10255 --platform linux/amd64 --detach mcr.microsoft.com/cosmosdb/linux/azure-cosmos-emulator:latest

This command gets me this:

```
<jemalloc>: MADV_DONTNEED does not work (memset will be used instead)
<jemalloc>: (This is the expected behaviour if you are running under QEMU)
./cosmosdb-emulator: ERROR: Invalid mapping of address 0x4005355000 in reserved address space below 0x400000000000. Possible causes:
1) the process (itself, or via a wrapper) starts-up its own running environment sets the stack size limit to unlimited via syscall setrlimit(2);
2) the process (itself, or via a wrapper) adjusts its own execution domain and flag the system its legacy personality via syscall personality(2);
3) sysadmin deliberately sets the system to run on legacy VA layout mode by adjusting a sysctl knob vm.legacy_va_layout.
```

I'm running Apple M3 Pro, Sonoma 14.7.1, Colima (started with the --arch amd flag)
